DPL Analysis: introduce filter selection caching

aalkin opened the pull request:
* Renamed `FilteredPolicy` to `FilteredBase` to better describe its role.
* `Filtered<>` now uses span to handle the selection, backed either by internal cache or external buffer. 
* Framework-managed `Filtered` and `Partition` now have their selections and gandiva filters cached by the framework per data frame, improving performance in cases where they are used in several `process()` functions
* Derived `Filtered` objects (sum or intersection) switch to internal cache automatically
* Combinations with filters will use internal caches
* Fixed bug in Combinations where temporary `Filtered` objects were not owned

aalkin committed:
Introduce trees/filters/selections caching for Filtered/Partition

aalkin committed:
Handle filter combinations; update test_ASoA.cxx

aalkin committed:
fixup! Handle filter combinations; update test_ASoA.cxx

alibuild committed:
Adapt to new FairLogger API

aalkin committed:
Merge pull request #7 from alibuild/alibot-refactoring-7726

Please consider the refactoring changes to #7726

ktf reviewed (changes requested):
Seems to break the tests.

aalkin committed:
Fix remaining memory issues

aalkin commented:
Now it passes tests, but breaks tasks in O2Physics that use combinations, investigating.

aalkin committed:
make sure combination policies do have internal cache for temporary tables

aalkin commented:
@ktf the latest change fixes the remaining issues.

ktf commented:
Seems to fail systematically on hyperloop https://alimonitor.cern.ch/train-workdir/tests/0001/00016096/stdout.log

aalkin committed:
add missing references/mutable

aalkin commented:
@ktf should be working now: apparently when I removed `const` in various places, I've also removed `&` which lead to selection cache not being updated. Seeing as everything else was already working, could you test this on hyperloop without waiting for CI?

ktf commented:
Tested on hyperloop. Merging.
